### PR TITLE
Harry/5019

### PR DESCRIPTION
**Pull request by hwh33**
This replaces `filters.Context` with `filters.ConnectionState`.  The former implemented the `context.Context` interface and was used as a connection-scoped context.  The main issue with this is that contexts are generally expected to be request-scoped and these long-lived contexts created performance issues.  For more details, see https://github.com/getlantern/lantern-internal/issues/5019.

An additional issue IMO is that the connection-scoped contexts were used as a bit of a crutch to pass data around function boundaries.  Hopefully the `ConnectionState` type makes the logic in this package a little more straight-forward.  That said, there are some cases where `ConnectionState` makes for an awkward replacement.  I tried to refactor so that use of `ConnectionState` felt natural, but also tried to avoid making too many logic changes.  I welcome any suggestions on this front.

### CODE REVIEW

File: proxy.go [hunk 1]
```diff
@@ -40,11 +40,11 @@ type DialFunc func(ctx context.Context, isCONNECT bool, network, addr string) (c
 type Proxy interface {
 	// Handle handles a single connection, with in specified separately in case
 	// there's a buffered reader or something of that sort in use.
-	Handle(ctx context.Context, in io.Reader, conn net.Conn) error
+	Handle(in io.Reader, conn net.Conn) error
 
 	// Connect opens a CONNECT tunnel to the origin without requiring a CONNECT
 	// request to first be sent on conn. It will not reply with CONNECT OK.
-	Connect(ctx context.Context, in io.Reader, conn net.Conn, origin string) error
+	Connect(in io.Reader, conn net.Conn, origin string) error
```
Comment (hwh33) [inline proxy.go:47]: (This comment applies to the `Handle` method above as well.)

We can remove the use of `ctx` here because we have replaced all use of it with logic based on passing `filters.ConnectionState` around.  There is another common use-case for `context.Context` parameters which is to implement cancellation and deadlines.  However (1) deadlines were barely handled and cancellations not at all and (2) I didn't find any code in any dependent projects which relied on these methods to handle cancellation and deadlines.

File: server.go [hunk 2]
```diff
@@ -14,6 +13,10 @@ func (proxy *proxy) Serve(l net.Listener) error {
 		if err != nil {
 			return errors.New("Unable to accept: %v", err)
 		}
-		go proxy.Handle(context.Background(), conn, conn)
+		go func() {
+			if err := proxy.Handle(conn, conn); err != nil {
+				log.Errorf("Error handling connection %v:", err)
```
Comment (hwh33) [inline server.go:18]: Drive-by: I noticed the error was ignored

**Comment by hwh33** [comment]
@ox I added you as a reviewer as I believe you are the most familiar with this logic and some of it is subtle.  Feel free to nominate someone else though if needed

File: filters/connection_state.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	upstreamAddr         string
+
+	requestNumber int
+
+	// Is this part of a MITM'd connection?
+	mitming bool
+
+	// Used by proxy.RequestAware connections.
+	requestAwareRequest  *http.Request
+	requestAwareUpstream net.Conn
```
[the remark this answers is not in the document]
Comment (oxtoacart) [reply filters/connection_state.go:26]: This is only a problem because the ConnectionState is strongly typed. If the ConnectionState where just a map (which is more like what context.Context was), the consumers of that map could handle the keys.

```go
type ConnectionState map[string]interface{}
```

If you want a nice interface, you could use a typedef in the `proxy` package like this:

```go
type ProxyConnectionState filters.ConnectionState

func (cs ProxyConnectionState) RequestAwareRequest() *http.Request {
    req, found := cs['req']
    if !found {
        return nil 
    }
    return req.(*http.Request)
}
```

Yes the type casting costs a little something, but I don't think we're using this stuff in a tight loop anywhere so I wouldn't be too concerned about it.

[the remark this answers is not in the document]
Comment (oxtoacart) [reply filters/connection_state.go:26]: That said, `filters` is only meaningful in the context of `proxy`, so the circularity you point out doesn't bother me all that much.

File: go.mod [hunk 4]
```diff
@@ -1,4 +1,4 @@
-module github.com/getlantern/proxy
+module github.com/getlantern/proxy/v2
```
[the remark this answers is not in the document]
Comment (oxtoacart) [reply go.mod:1]: I'm fine with this.

File: proxy.go (hunk 1 above)
Comment (oxtoacart) [reply proxy.go:47]: I see that we use it in flashlight in `client.go` in our socks5 config:

```
conf := &socks5.Config{
		HandleConnect: func(ctx context.Context, conn net.Conn, req *socks5.Request, replySuccess func(boundAddr net.Addr) error, replyError func(err error) error) error {
			op, ctx := ops.BeginWithNewBeam("proxy", ctx)
			defer op.End()

			host := fmt.Sprintf("%v:%v", req.DestAddr.IP, req.DestAddr.Port)
			addr, err := client.reverseDNS(host)
			if err != nil {
				return op.FailIf(log.Errorf("Error performing reverseDNS for %v: %v", host, err))
			}
			errOnReply := replySuccess(nil)
			if errOnReply != nil {
				return op.FailIf(log.Errorf("Unable to reply success to SOCKS5 client: %v", errOnReply))
			}
			return op.FailIf(client.proxy.Connect(ctx, req.BufConn, conn, addr))
		},
	}
```

And I see that we use it in dialing [here](https://github.com/getlantern/proxy/pull/50/files#diff-7a0eda0d43627b11b85f773a1b2cab7fff6b7b20e8dc65beb8d638aabc3c6c3bL95), so I don't think it's safe to discard it.

File: filters/connection_state.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hwh33) [reply filters/connection_state.go:26]: I considered a map actually, but I think that creates the same kind of issues with subtle logic that we had with the context (for the same reason - it's just a bag-o-values).  With a struct, we can see which methods are called where.

> That said, filters is only meaningful in the context of proxy, so the circularity you point out doesn't bother me all that much.

Yeah, that fact that the circular bit is pretty hidden away by the API makes me more okay with it too (basically callers don't need to go back and forth between the packages, they can just implement the `RequestAware` and `ResponseAware` interfaces).

File: proxy_connect.go [hunk 5]
```diff
@@ -146,25 +155,18 @@ func addDialDeadlineIfNecessary(ctx context.Context, req *http.Request) (context
 func noopCancel() {
 }
 
-func respondOK(resp *http.Response, req *http.Request, ctx filters.Context) (*http.Response, filters.Context) {
```
[the remark this answers is not in the document]
Comment (hwh33) [reply proxy_connect.go:149]: In the old code, you can examine references to `ctxKeyNoRespondOkay` to see that we only suppress the OK response if the connection comes in through `proxy.Connect` ([here](https://github.com/getlantern/proxy/blob/master/proxy_connect.go#L161)).  Rather than making this choice and tracking it as state on an object, it seems more straight-forward to pass the choice as a parameter through the call-stack.

File: filters/middleware.go [hunk 6]
```diff
@@ -2,29 +2,24 @@ package filters
...
 func Intercept(handler http.Handler, filter Filter) http.Handler {
-	var conn net.Conn
 	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
-		getDownstream := func() net.Conn {
-			if conn == nil {
-				conn, _, _ = resp.(http.Hijacker).Hijack()
-			}
-			return conn
+		var cs *ConnectionState
+		if hj, ok := resp.(http.Hijacker); ok {
```
[the remark this answers is not in the document]
Comment (hwh33) [reply filters/middleware.go:13]: Hmm good catch. Looking into this

**Comment by hwh33** [comment]
> Finished my first pass. I want to take an even closer look a little later this evening.

Thank you!  There's some subtle and difficult-to-follow logic here, so I definitely appreciate the close inspection!

File: proxy_connect.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (oxtoacart) [reply proxy_connect.go:149]: That's fine, but you could still keep lines 152-154 which are repeated.

File: proxy.go (hunk 1 above)
Comment (oxtoacart) [reply proxy.go:47]: Yes, the upstream dial timing out is the main case we're concerned about, as in office that can in fact hang.

File: filters/middleware.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (hwh33) [reply filters/middleware.go]: The only use of `filters.Intercept` is [here](https://github.com/getlantern/http-proxy-lantern/blob/a1775b7923b5bc6c551321dc70fc4f1001db54c3/http_proxy.go#L419) in http-proxy-lantern.  However, it's still difficult to tell what would happen if hijacking were not deferred.  I decided to be safe and delay hijacking as you suggested: https://github.com/getlantern/proxy/pull/50/commits/d617367507d57a9076539462082aeacfc44ef8dc

[the remark this answers is not in the document]
Comment (hwh33) [reply filters/middleware.go]: Thanks for catching this!  That was not an intentional change and it would have been a real PITA to debug if it did break something

**Comment by oxtoacart** [comment]
Thinking about this some more, I think the reason for wrapping context.Context in the first place is that the context may somewhere get passed to some Go code and eventually get passed back to us. As originally implemented, that would allow our values to pass through, but with your changes it wouldn't. Can you please make sure that we don't need to support that usage anywhere in flashlight or in http-proxy or http-proxy-lantern?

Also, it would be nice to have a benchmark for the before/after to make sure this makes as big of a difference as we hope.

**Comment by hwh33** [comment]
> If you like the lower risk of my PR that doesn't actually change the API and doesn't require any code changes outside of the proxy project, feel free to use that instead :)

Oh haha, I deferred to you around the same time [here](https://github.com/getlantern/proxy/pull/51#issuecomment-937433939).  Want to hop on a call tomorrow and talk a little about both approaches?  I wouldn't mind taking the night to sleep on it.

File: proxy.go (hunk 1 above)
Comment (oxtoacart) [reply proxy.go:47]: I like adding a `dialTimeout` or something like that.

**Comment by hwh33** [comment]
> I'd be curious what paths lead specifically to bad performance. I think the major issue here is actually memory usage, since that linked list of Contexts could grow to be quite large.

@benjojo found a few proxies with random CPU spikes:
image

and crazy call-stack flame graphs:
image

These proxies were spending a lot of CPU time unwinding the context.Value chain:
image

and had some crazy-looking call-stacks, explained only as a massive unwinding of the context.Value chain:
```
1 @ 0x4a9f68 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9efd 0x4a9e
#       0x4a9f67        context.(*valueCtx).Value+0xc7  /usr/local/go/src/context/context.go:558
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
#       0x4a9efc        context.(*valueCtx).Value+0x5c  /usr/local/go/src/context/context.go:562
```

We connected the dots when we found the `context.WithValue` implementation and noticed that contexts were connection-scoped.  This was corroborated by the fact that the proxies experiencing the issue were all on multiplex tracks (which should have longer connection lifetimes).

I'm not sure that memory consumption was _not_ an issue on these machines.  We were just focused on CPU usage spikes.

**Comment by hwh33** [comment]
Okay, sure.

> Caller needs to be able to supply a dialTimeout for the upstream dial

Okay, this can be managed.  Thanks for bringing attention to that earlier.

> The benchmark from my PR would be handy to add here just to illustrate what performance issue this solves.

Sure, we can add that.  Did [this](https://github.com/getlantern/proxy/pull/51#discussion_r723842914) make sense or was I just missing the goal of your benchmark?

**Comment by hwh33** [comment]
@oxtoacart I originally tried a dial timeout as discussed, but ended up adding a context back in, specific to the dial.  As you suggested, it makes the most sense for this to apply only to the dial, not the entire function call.  I specified this in the parameter name and documentation.  Changes here: https://github.com/getlantern/proxy/pull/50/commits/571214ef7a70cd3ef77249b05329a473bdb4027b

I also incorporated this change into the flashlight PR here: https://github.com/getlantern/flashlight/pull/1148/commits/03be5fbce76975b8ed1ee0fbb625bafcca8297b4

I know we discussed adding a benchmark, but I'm actually not sure what to benchmark.  The comparison would now be between `context.Context.Value` and field access (e.g. `ConnectionState.RequestNumber()`).  I think it's clear that accessing fields on a struct is always going to scale.  What do you think?

**Review by oxtoacart — APPROVED** [review]
LGTM!

**Comment by hwh33** [comment]
Thanks @oxtoacart!
